tests/scanner/scanners/external_project_access_scanner_test.py

```diff
@@ -0,0 +1,301 @@
+        """Setup"""
+        self.global_configs = dict()
+        self.scanner_configs = (
+            dict(output_path='gs://test-forseti-dev/scanner/output',
+                 rules_path=__file__,
+                 scanners=[dict(name='external_project_access', enabled=True)]))
+
+        mock_invnetory_config = mock.MagicMock()
+        mock_invnetory_config.get_root_resource_id.return_value = (
+            'organizations/567890')
+
+        mock_service_config = mock.MagicMock()
+        mock_service_config.get_inventory_config.return_value = (
+            mock_invnetory_config)
+
+        self.service_config = mock_service_config
+        self.model_name = 'TestModel'
+        self.snapshot_timestamp = datetime.now().strftime('%Y%m%dT%H%M%SZ')
+        self.rules = get_datafile_path(
+            __file__, 'external_project_access_test_rules_1.yaml')
+
+        self.stash_email_method = epas.get_user_emails
+
+    def test_mock_emails(self):
+        """Test obtaining e-mails from mock"""
+        epas.get_user_emails = mock.MagicMock(return_value=TEST_EMAILS)
+        self.assertEqual(len(epas.get_user_emails(None)), 2)
+
+    #pylint: disable=W0212
+    def test_retrieve(self):
+        """Test retrieving project ancestry data"""
+        list_to_generator = lambda x: (n for n in x)
+
+        epas.get_user_emails = mock.MagicMock(return_value=TEST_EMAILS)
+        epas.extract_project_ids = mock.MagicMock(
+            return_value=['project1', 'project2', 'project3'])
+
+        api_helpers.get_delegated_credential = mock.MagicMock()
+
+        crm.get_projects = mock.MagicMock(
+            return_value=list_to_generator(
+                [fake_crm_responses.FAKE_PROJECTS_API_RESPONSE1]))
+
+        crm.CloudResourceManagerClient.get_project_ancestry = mock.MagicMock(
+            return_value=json.loads(
+                fake_crm_responses.GET_PROJECT_ANCESTRY_RESPONSE)['ancestor'])
+
+        scanner = epas.ExternalProjectAccessScanner(self.global_configs,
+                                                    self.scanner_configs,
+                                                    self.service_config,
+                                                    self.model_name,
+                                                    self.snapshot_timestamp,
+                                                    self.rules)
+        scanner.rules_engine.build_rule_book(
+            self.service_config.get_inventory_config())
+
+        user_ancestries = scanner._retrieve()
+
+        self.assertIn('user1@example.com', user_ancestries.keys())
+        self.assertTrue(isinstance(user_ancestries['user1@example.com'], list))
+        self.assertTrue(
+            isinstance(user_ancestries['user1@example.com'][0], list))
+        self.assertTrue(
+            isinstance(user_ancestries['user1@example.com'][0][0], Project))
+
+    def test_find_violations_bad_folder(self):
+        """Test finding no violations with a folder as a parent"""
```

Comment:
You should add `bad` here in the description.
```
"""Test finding no violations with a bad folder as a parent"""
```

Reply:
modified as suggested